src/jit/codegenxarch.cpp

```diff
@@ -7872,25 +7919,41 @@ CodeGen::genIntrinsic(GenTreePtr treeNode)
 // Return value:
 //    The number of the base variable.
 //
+// Note:
+//    If tail call the outgoing args are placed in the callers outgoing stack space.
+//    Otherwise, they go in the outgoing area on the current frame.
+//    
+//    On Windows the caller always creates slots (homing space) in its frame for the 
+//    first 4 arguments of a calee (register passed args). So, the baseVarNum is always 0.
+//    For System V systems there is no such calling convention requirement, and the code needs to find
+//    the first stack passed argument from the caller. This is done by iterating over 
+//    all the lvParam variables and finding the first with lvArgReg equals to REG_STK.
+//    
 unsigned
 CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
 {
     assert(treeNode->OperGet() == GT_PUTARG_STK);
 
-    unsigned baseVarNum;
+    unsigned baseVarNum = 0;
+
 #if FEATURE_FASTTAILCALL
     bool putInIncomingArgArea = treeNode->AsPutArgStk()->putInIncomingArgArea;
 #else
     const bool putInIncomingArgArea = false;
 #endif
+
     // Whether to setup stk arg in incoming or out-going arg area?
     // Fast tail calls implemented as epilog+jmp = stk arg is setup in incoming arg area.
     // All other calls - stk arg is setup in out-going arg area.
     if (putInIncomingArgArea)
     {
-        // The first baseVarNum is guaranteed to be the first incoming arg of the method being compiled.
-        // See lvaInitTypeRef() for the order in which lvaTable entries are initialized.
-        baseVarNum = 0;
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        // See the note in the function header re: finding the first stack passed argument.
+        baseVarNum = genGetFirstArgWithStackSlot();
+#endif // !FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        assert(baseVarNum != BAD_VAR_NUM);
+
```

Comment:
That is strange! because on Linux varDesc corresponds to an arg that is passed on STK.  varDsc->lvIsRegArg has to be false, no?

It makes me suspect that the way we are finding the first arg with stack slot seems to be incorrect.  We need to find out where is the mismatch.

Though i thought this assert is not useful, it has some value on second though.  We need to change the assert on Linux appropriately.